src/include/duckdb/storage/compression/alp/algorithm/alp.hpp

```diff
@@ -0,0 +1,397 @@
+	}
+
+	/*
+	 * Find the best combinations of factor-exponent from each vector sampled from a rowgroup
+	 */
+	static void FindTopKCombinations(vector<vector<T>> vectors_sampled, State &state) {
+
+		// We use a 'pair' to hash it easily
+		state.best_k_combinations.clear();
+		map<pair<int8_t, int8_t>, int32_t> best_k_combinations_hash;
+
+		// For each vector sampled
+		for (auto &sampled_vector : vectors_sampled) {
+			idx_t n_samples = sampled_vector.size();
+			uint8_t best_factor = AlpTypedConstants<T>::MAX_EXPONENT;
+			uint8_t best_exponent = AlpTypedConstants<T>::MAX_EXPONENT;
+
+			//! We start our optimization with the worst possible total bits obtained from compression
+			idx_t best_total_bits = (n_samples * (EXACT_TYPE_BITSIZE + AlpConstants::EXCEPTION_POSITION_SIZE * 8)) +
+			                        (n_samples * EXACT_TYPE_BITSIZE);
+
+			//! We try all combinations in search for the one which minimize the compression size
+			for (int8_t exp_idx = AlpTypedConstants<T>::MAX_EXPONENT; exp_idx >= 0; exp_idx--) {
+				for (int8_t factor_idx = exp_idx; factor_idx >= 0; factor_idx--) {
+					idx_t exceptions_count = 0;
+					idx_t non_exceptions_count = 0;
+					uint32_t estimated_bits_per_value = 0;
+					uint64_t estimated_compression_size = 0;
+					int64_t max_encoded_value = NumericLimits<int64_t>::Minimum();
+					int64_t min_encoded_value = NumericLimits<int64_t>::Maximum();
+
+					for (idx_t sample_idx = 0; sample_idx < n_samples; sample_idx++) {
+						T value = sampled_vector[sample_idx];
+
+						T decoded_value;
+						T tmp_encoded_value;
+						int64_t encoded_value;
+
+						tmp_encoded_value =
+						    value * AlpTypedConstants<T>::EXP_ARR[exp_idx] * AlpTypedConstants<T>::FRAC_ARR[factor_idx];
+						encoded_value = NumberToInt64(tmp_encoded_value);
+
+						//! The cast to T is needed to prevent a signed integer overflow
+						decoded_value = static_cast<T>(encoded_value) * AlpConstants::FACT_ARR[factor_idx] *
+						                AlpTypedConstants<T>::FRAC_ARR[exp_idx];
+						if (decoded_value == value) {
+							non_exceptions_count++;
+							max_encoded_value = MaxValue(encoded_value, max_encoded_value);
+							min_encoded_value = MinValue(encoded_value, min_encoded_value);
+							continue;
+						}
+						exceptions_count++;
+					}
+
+					if (non_exceptions_count < 2) { // We skip combinations which yields to almost all exceptions
+						continue;
+					}
+					// Evaluate factor/exponent compression size (we optimize for FOR)
+
+					uint64_t delta =
+					    (static_cast<uint64_t>(max_encoded_value) - static_cast<uint64_t>(min_encoded_value));
+					estimated_bits_per_value = std::ceil(std::log2(delta + 1));
+					estimated_compression_size += n_samples * estimated_bits_per_value;
+					estimated_compression_size +=
+					    exceptions_count * (EXACT_TYPE_BITSIZE + (AlpConstants::EXCEPTION_POSITION_SIZE * 8));
+
+					if ((estimated_compression_size < best_total_bits) ||
```

Comment:
Not entirely. Here we compare the estimated compression size of the current combination vs. the best combination (so far). In `CompareALPCombinations`, we mainly compare the number of appearances of the combination within sampled vectors. 

However, I do agree that this may cause confusion to the reader. 
We have merged both logics by adding an additional field to the `AlpCombination` struct: `estimated_compression_size`, to use it in the comparator function, and to use the comparator function every time we want to check if a combination is better than another one